IbFlexReader/IbFlexReader/Reader.cs

```diff
@@ -2,92 +2,215 @@
+        /// <param name="retry">if GetStatement returns warning 1019 (... Please try again shortly), retry defines how many times it will be tried again</param>
         /// <returns></returns>
-        public async Task<FlexResult> GetByApi(string token, string queryId, string dumpFile = null)
-        {
-            using (var client = new HttpClient())
-            {
-                var uri = new Uri($"https://gdcdyn.interactivebrokers.com/Universal/servlet/FlexStatementService.SendRequest?t={token}&q={queryId}&v=3");
-
-                string referenceCode;
-                using (var sendRequestResult = await client.PostAsync(uri, null))
-                using (var stream = await sendRequestResult.Content.ReadAsStreamAsync())
-                {
-                    if (dumpFile != null) { Utils.DumpFileWriter.DumpSendRequest(stream, dumpFile); }
-
-                    var response = Deserializer.Deserialize<XmlFlexStatementResponse, FlexStatementResponse>(stream, out var errorObjects);
-
-                    if (response.Status != "Success")
-                    {
-                        return new FlexResult
-                        {
-                            ErrorCode = response.ErrorCode,
-                            ErrorMessage = response.ErrorMessage
-                        };
-                    }
-
-                    referenceCode = response.ReferenceCode.Value.ToString();
-                }
-
-
-                uri = new Uri($"https://gdcdyn.interactivebrokers.com/Universal/servlet/FlexStatementService.GetStatement?q={referenceCode}&t={token}&v=3");
-
-                using (var getStatementResult = await client.PostAsync(uri, null))
-                using (var stream = await getStatementResult.Content.ReadAsStreamAsync())
+        public async Task<FlexResult> GetByApi(string token, string queryId, string dumpFile = null, int? retry = null)
+		{
+			using (var client = new HttpClient())
+			{
+				FlexResult requestResult = await SendRequest(client, token, queryId, dumpFile);
+
+				if (!requestResult.IsSuccess)
+				{
+					return requestResult;
+				}
+
+				string referenceCode = requestResult.ErrorMessage;
+
+				int retryCounter = 0;
+				if (retry.HasValue && retry.Value > 0) retryCounter = retry.Value;
+
+				FlexResult result = null;
+
+				bool tryAgain = true;
+				while (tryAgain)
+				{
+					result = await GetStatement(client, token, referenceCode, dumpFile);
+
+					if (!result.IsSuccess && result.ErrorCode.HasValue && result.ErrorCode.Value == 1019)
+					{
+                        // warning is: 'Statement generation in progress. Please try again shortly. '
+
+                        System.Diagnostics.Debug.Print("IbFlexReader.Reader.GetByApi: Retry");
+
+						// retry getting data?
+						if (retryCounter > 0)
+						{
+							// yes
+							retryCounter--;
+                            // wait a little bit 
+							System.Threading.Thread.Sleep(2000);
```

Comment:
Please use `await Task.Delay(2000)` as this is the better way of delaying some code in asynchronous methods. `Task.Delay()` will release the current thread while `Thread.Sleep` will block the thread.

Reply:
Done